source/tutorials/nix-language.md

```diff
@@ -0,0 +1,1766 @@
+
+- Compound data types and, in the broader sense, functions
+
+  to produce and transform complex data
+
+- Name assignment
+
+  to manipulate complex data as units
+
+The language is *pure*, that is, its evaluation does not observe or interact with the outside world – with one notable exception:
+reading files, to capture what build tasks will operate on.
+
+There is nothing else to it.
+What may look complicated comes not from the language, but from how it is used.
+
+### Notes on syntax
+
+This tutorial does not go into the specifics of syntactical rules, but rather puts emphasis on recurrent patterns as observed in practical code and what they mean.
+
+The most important heuristics to avoid confusion are about white space and parentheses, since in that regard, the Nix language may differ from languages you are used to.
+
+(Don't worry if you don't understand the examples here, they are only for illustration.)
+
+1. White space is used to delimit [lexical tokens], where required.
+
+   It is otherwise insignificant.
+   Line breaks, indentation, and additional spaces are for readers' convenience.
+
+   The following are equivalent:
+
+   ```nix
+   let
+     x = 1;
+     y = 2;
+   in x + y
+   ```
+
+       3
+
+   ```nix
+   let x=1;y=2;in x+y
+   ```
+
+       3
+
+[lexical tokens]: https://en.m.wikipedia.org/wiki/Lexical_analysis#Token
+
+2. Parentheses are used to force precedence.[^inherit-parens]
+
+   The following are different:
+
+   ```nix
+   let
+     f = x: x + 1;
+     a = 1;
+   in [ (f a) ]
+   ```
+
+       2
+
+   ```nix
+   let
+     f = x: x + 1;
+     a = 1;
+   in [ f a ]
+   ```
+
+       [ <LAMBDA> 1 ]
+
+[^inherit-parens]: [`inherit`](inherit) is the only exception where parentheses are a syntactical requirement.
+
+(names-values)=
+## Names and values
+
+Values in the Nix language can be primitive data types, lists, attribute sets, and functions.
+
+We show primitive data types and lists as examples in the context of [attribute sets](attrset).
+Later in this section we cover special features of character strings: [antiquotation](antiquotation), [file system paths](file-system-paths), and [indented strings](indented-strings).
+We deal with [functions](functions) separately.
+
+There are two ways to assign names to values in Nix: [attribute sets](attrset) and [`let` expressions](let).
+
+Assignments are denoted by a single equal sign (`=`).
+
+Whenever you see an equal sign (`=`) in Nix language code:
+- On its left is the assigned name.
+- On its right is the value, delimited by a semicolon (`;`).
+
+(attrset)=
+### Attribute set `{ ... }`
+
+An attribute set is a collection of name-value-pairs, where names must be unique.
+
+The following example shows all primitive data types, lists, and attribute sets.
+
+:::{note}
+If you are familiar with JSON, imagine the Nix language as *JSON with functions*.
+
+Nix language data types *without functions* work just like their counterparts in JSON and look very similar.
```

Comment:
This is not true. It's fine to say *think about* Nix as JSON plus functions, but this is *only* an analogy. Specifically:

* Strings in C++ Nix are byte sequences and not Unicode-encoded in some shape or form as in JSON.
* Integers in Nix are 64 bit integers and not represented as floating point numbers as they are in JSON.
* Floats overall behave weirdly in Nix to the point that you should warn about them, i.e. they get truncated when serializing them in many situations.
* Due to the nature of Nix's strings, not every legal Nix attribute name is a legal JSON object key.

Reply:
I think the formulation "imagine the Nix language as" is somehow similar to saying "think about Nix as", isn't it?

Reply:
Yes, but the claim in the second sentence is stronger than that and wrong, as I've pointed out. “Nix language data types *without functions*” does *not* work just like their counterparts in JSON.

Reply:
I feel that it is at least implied. With tutorials as the first contact point, one should be careful to avoid creating wrong impressions in readers – even by accident. FWIW my point is not that these differences *must* be explained, but rather that we should be careful to word it in a way that we don't imply they don't exist.

---

I'm also not sure these edges are that far out.

The integer thing is a relatively well known JSON pitfall – Nix is not thorny here as it behaves like you'd expect.

```console
> echo 34343499993845656749 | jq
34343499993845658000
```

The float situation can be discovered quite easily. E.g.:

```console
nix-repl> 1.000004 
1
nix-repl> builtins.toJSON 1.000004
"1"
```

String indexing, I'd say is a very basic topic that should be covered in a tutorial, since the internal representation is apparent in this, it is relevant. Indexing into non-ASCII strings will unexpectedly  e.g. `builtins.substring 0 1 "καλημερα"` will return in some garbage being returned. E.g. “Nix strings are similar to C strings” is a statement like the one about JSON data, but more accurate.